datalad/distribution/install.py

```diff
@@ -310,90 +162,144 @@ def __call__(
-                        return installed_items
-            else:
-                path = path[0]
+                        lgr.warning("%s was not installed", ds_)
+
+            return installed_items[0] \
+                if len(installed_items) == 1 else installed_items
+
+        if source and path and len(path) > 1:
+            raise ValueError(
+                "install needs a single PATH when source is provided.  "
+                "Was given mutliple PATHs: %s" % str(path))
+
+        # parameter constraints:
+        if not source:
+            raise InsufficientArgumentsError(
+                "a `source` is required for installation")
+
+        # code below deals with a single path only
+        path = path[0] if path else None
+
+        if source == path:
+            # even if they turn out to be identical after resolving symlinks
+            # and more sophisticated witchcraft, it would still happily say
+            # "it appears to be already installed", so we just catch an
+            # obviously pointless input combination
+            raise ValueError(
+                "installation `source` and destination `path` are identical. "
+                "If you are trying to add a subdataset simply use `save` %s".format(
+                    path))
 
-        # now the 'usual' flow with single `path`argument:
-        # shortcut
-        ds = dataset
+        installed_items = []
 
-        _install_into_ds = False  # default
         # did we explicitly get a dataset to install into?
         # if we got a dataset, path will be resolved against it.
         # Otherwise path will be resolved first.
-
-        if ds is not None:
-            _install_into_ds = True
-            if not isinstance(ds, Dataset):
-                ds = Dataset(ds)
-            if not ds.is_installed():
-                # TODO: possible magic: ds.install() for known subdataset 'ds'
-                raise ValueError("{0} needs to be installed in order to "
-                                 "install something into it.".format(ds))
+        ds = None
+        if dataset is not None:
+            ds = require_dataset(dataset, check_installed=True,
+                                 purpose='installation')
             handle_dirty_dataset(ds, if_dirty)
 
         # resolve the target location (if local) against the provided dataset
         # or CWD:
         if path is not None:
             # Should work out just fine for regular paths, so no additional
             # conditioning is necessary
-            path_ri = RI(path)
+            try:
+                path_ri = RI(path)
+            except Exception as e:
+                raise ValueError(
+                    "invalid path argument {}: ({})".format(path, exc_str(e)))
             try:
                 # Wouldn't work for SSHRI ATM, see TODO within SSHRI
-                path = resolve_path(path_ri.localpath, ds)
+                path = resolve_path(path_ri.localpath, dataset)
```

Comment:
here it is not `me` -- it is `us` ;)  So let's analyze the code:  `dataset` is the one passed explicitly, and `ds` is the one handled (not really resolved much, but saved etc) IFF `dataset` was passed explicitly.  So as far as I see it -- it would be pretty much the same sh.t.  I added an assert on ds == dataset and didn't see any failure when running test_{install,get}.py .  So might be just a left over and could be either but I am reluctant to change it and will leave it to @mih since there might have been a higher thought